File: src/lib/finding-grouping.ts
import { normalizeToolSource } from '@/lib/catalog-tool-index';
import type { Finding, Severity } from '@/lib/secops-api';
import { fixTextEncoding } from '@/lib/text-encoding';

const PLUGIN_ID_RE = /Plugin ID:\s*(\d+)/i;
const HOST_RE = /^Host:\s*(.+)$/im;
const PORT_RE = /^Puerto:\s*(.+)$/im;

export const SALIDA_PREFIX_TEMPLATE = '-----[Salida correspondiente a: {componente} ]-----';

export function normalizeGroupingTitle(titulo: string): string {
  return (titulo || '').trim().toLowerCase();
}

export function extractNessusPluginId(raw: string | null | undefined): string | null {
  if (!raw) return null;
  const m = raw.match(PLUGIN_ID_RE);
  return m?.[1] ?? null;
}

export function resolveFindingComponente(finding: Finding): string {
  const direct = (finding.componente_afectado || '').trim();
  if (direct) return direct;
  const raw = finding.raw_tool_output || '';
  const host = raw.match(HOST_RE)?.[1]?.trim();
  if (!host) return '';
  const port = raw.match(PORT_RE)?.[1]?.trim();
  if (port && port !== '0' && port.toLowerCase() !== 'none') {
    return `${host}:${port}`;
  }
  return host;
}

function resolveFindingToolIdentity(finding: Finding): { source: string; id: string } {
  const id = (finding.tool_vuln_id || '').trim();
  if (id) {
    return { source: normalizeToolSource(finding.tool_source), id };
  }
  const plugin = extractNessusPluginId(finding.raw_tool_output);
  if (plugin) return { source: 'nessus', id: plugin };
  const title = (finding.titulo || '').trim();
  return { source: 'manual', id: title };
}

/** Clave de tipo de vulnerabilidad (sin severidad ni activo) — una fila en revisión de catálogo. */
export function findingTypeKey(finding: Finding): string {
  const { source, id } = resolveFindingToolIdentity(finding);
  if (id && source !== 'manual') return `tool:${source}:${id}`;
  const plugin = extractNessusPluginId(finding.raw_tool_output);
  if (plugin) return `plugin:${plugin}`;
  return `title:${normalizeGroupingTitle(finding.titulo)}`;
}

/** Clave de agrupación (mismo plugin/identificador + severidad). */
export function findingGroupKey(finding: Finding): string {
  const { source, id } = resolveFindingToolIdentity(finding);
  if (id && source !== 'manual') return `tool:${source}:${id}:${finding.severidad}`;
  const plugin = extractNessusPluginId(finding.raw_tool_output);
  if (plugin) return `plugin:${plugin}:${finding.severidad}`;
  return `title:${normalizeGroupingTitle(finding.titulo)}:${finding.severidad}`;
}

function uniqueLines(values: string[]): string[] {
  const seen = new Set<string>();
  const out: string[] = [];
  for (const v of values) {
    const t = v.trim();
    if (!t || seen.has(t)) continue;
    seen.add(t);
    out.push(t);
  }
  return out;
}

const PLUGIN_OUTPUT_RE = /plugin\s+output\s*:\s*/i;
const NESSUS_HEADER_LINE_RE =
  /^(?:\[Nessus CSV\]|Plugin ID:|Host:|Puerto:|Protocol:|CVE:|Synopsis:)/i;

/** Solo Plugin Output (sin metadatos Nessus de auto-ingesta). */
export function extractPluginOutputText(raw: string | null | undefined): string {
  const t = (raw || '').trim();
  if (!t) return '';
  const match = PLUGIN_OUTPUT_RE.exec(t);
  if (match) return t.slice(match.index + match[0].length).trim();
  if (t.toLowerCase().includes('[nessus csv]')) {
    const body: string[] = [];
    let pastHeaders = false;
    for (const line of t.split('\n')) {
      const stripped = line.trim();
      if (!stripped) {
        if (body.length) pastHeaders = true;
        continue;
      }
      if (!pastHeaders && NESSUS_HEADER_LINE_RE.test(stripped)) continue;
      pastHeaders = true;
      body.push(line);
    }
    return body.join('\n').trim();
  }
  return t;
}

export function personalizedSalida(componente: string, toolOutput: string): string {
  const cleaned = extractPluginOutputText(toolOutput);
  if (!cleaned) return '';
  const label = componente || 'N/A';
  return `${SALIDA_PREFIX_TEMPLATE.replace('{componente}', label)}\n${cleaned}`;
}

export type FindingGroup = {
  key: string;
  titulo: string;
  severidad: Severity;
  members: Finding[];
  componentes: string[];
  mergedComponentes: string;
  mergedSalidas: string;
};

export type VulnerabilityTypeGroup = {
  key: string;
  titulo: string;
  severidad: Severity;
  members: Finding[];
  instanceCount: number;
  toolLabel: string;
};

function maxSeverity(findings: Finding[]): Severity {
  const order: Severity[] = ['Critical', 'High', 'Medium', 'Low', 'Info'];
  let best = 99;
  let pick: Severity = 'Info';
  for (const f of findings) {
    const idx = order.indexOf(f.severidad);
    if (idx >= 0 && idx < best) {
      best = idx;
      pick = f.severidad;
    }
  }
  return pick;
}

function toolLabelForFinding(finding: Finding): string {
  const { source, id } = resolveFindingToolIdentity(finding);
  if (id) return `${source}:${id}`;
  const plugin = extractNessusPluginId(finding.raw_tool_output);
  if (plugin) return `nessus:${plugin}`;
  return 'manual';
}

/** Agrupa por tipo de vulnerabilidad (sin IP/puerto); una entrada por plugin/identificador. */
export function groupFindingsByVulnerabilityType(findings: Finding[]): VulnerabilityTypeGroup[] {
  const buckets = new Map<string, Finding[]>();
  const order: string[] = [];

  for (const f of findings) {
    const key = findingTypeKey(f);
    if (!buckets.has(key)) {
      buckets.set(key, []);
      order.push(key);
    }
    buckets.get(key)!.push(f);
  }

  return order.map((key) => {
    const members = buckets.get(key)!;
    const rep = members[0];
    return {
      key,
      titulo: (rep.titulo || '').trim() || 'Sin título',
      severidad: maxSeverity(members),
      members,
      instanceCount: members.length,
      toolLabel: toolLabelForFinding(rep),
    };
  });
}

export function groupFindingsForDisplay(findings: Finding[]): FindingGroup[] {
  const buckets = new Map<string, Finding[]>();
  const order: string[] = [];

  for (const f of findings) {
    const key = findingGroupKey(f);
    if (!buckets.has(key)) {
      buckets.set(key, []);
      order.push(key);
    }
    buckets.get(key)!.push(f);
  }

  return order.map((key) => {
    const members = buckets.get(key)!;
    const componentes = uniqueLines(members.map(resolveFindingComponente));
    const salidas = members
      .map((m) => personalizedSalida(resolveFindingComponente(m), m.raw_tool_output || ''))
      .filter(Boolean);
    return {
      key,
      titulo: fixTextEncoding(members[0].titulo),
      severidad: members[0].severidad,
      members,
      componentes,
      mergedComponentes: componentes.join('\n'),
      mergedSalidas: salidas.join('\n'),
    };
  });
}

/** Agrupa por ai_group_id; sin ID usa título normalizado como clave de respaldo. */
export function groupFindingsByAiGroup(findings: Finding[]): FindingGroup[] {
  const buckets = new Map<string, Finding[]>();
  const order: string[] = [];

  for (const f of findings) {
    const key = f.ai_group_id
      ? `ai:${f.ai_group_id}`
      : `title:${normalizeGroupingTitle(f.titulo)}`;
    if (!buckets.has(key)) {
      buckets.set(key, []);
      order.push(key);
    }
    buckets.get(key)!.push(f);
  }

  return order
    .map((key) => {
      const members = buckets.get(key)!;
      if (members.length < 2 && key.startsWith('title:')) return null;
      const componentes = uniqueLines(members.map(resolveFindingComponente));
      const salidas = members
        .map((m) => personalizedSalida(resolveFindingComponente(m), m.raw_tool_output || ''))
        .filter(Boolean);
      const rep = members[0];
      const label = rep.ai_group_id
        ? fixTextEncoding(rep.titulo)
        : `${fixTextEncoding(rep.titulo)} (título)`;
      return {
        key,
        titulo: label,
        severidad: maxSeverity(members),
        members,
        componentes,
        mergedComponentes: componentes.join('\n'),
        mergedSalidas: salidas.join('\n'),
      };
    })
    .filter((g): g is FindingGroup => g !== null && g.members.length > 0);
}

export function countAiGroups(findings: Finding[]): number {
  const keys = new Set<string>();
  for (const f of findings) {
    if (f.ai_group_id) keys.add(f.ai_group_id);
    else keys.add(`title:${normalizeGroupingTitle(f.titulo)}`);
  }
  return keys.size;
}

File: backend/app/services/finding_grouping.py
"""Agrupa hallazgos por título + severidad para reportes Word (Power Query / VBA)."""

from __future__ import annotations

import re
import uuid
from collections import OrderedDict
from dataclasses import dataclass, field
from typing import Optional

from app.models.core import Asset, Finding, Severity
from app.services.report_text_preprocess import preprocess_report_field, salida_herramienta_for_report
from app.services.catalog_tool_index import normalize_tool_source, resolve_finding_tool_identity
from app.services.text_encoding import extract_nessus_plugin_id

SALIDA_PREFIX_TEMPLATE = "-----[Salida correspondiente a: {componente} ]-----"
_HOST_RE = re.compile(r"^Host:\s*(.+)$", re.MULTILINE | re.IGNORECASE)
_PORT_RE = re.compile(r"^Puerto:\s*(.+)$", re.MULTILINE | re.IGNORECASE)


def normalize_grouping_title(titulo: str) -> str:
    """Clave de agrupación: trim + comparación sin distinguir mayúsculas."""
    return (titulo or "").strip().casefold()


def grouping_key(finding: Finding) -> tuple[str, ...]:
    """Índice herramienta (tipo origen + id) o título normalizado + severidad."""
    src, vid = resolve_finding_tool_identity(finding)
    if vid and normalize_tool_source(src) != "manual":
        return (
            "tool",
            normalize_tool_source(src),
            vid,
            finding.severidad.value if finding.severidad else "",
        )
    plugin = extract_nessus_plugin_id(finding.raw_tool_output)
    if plugin:
        return ("plugin", plugin, finding.severidad.value if finding.severidad else "")
    return ("title", normalize_grouping_title(finding.titulo), finding.severidad.value if finding.severidad else "")


def _host_from_raw(raw: str) -> str:
    host_match = _HOST_RE.search(raw)
    if not host_match:
        return ""
    host = host_match.group(1).strip()
    port_match = _PORT_RE.search(raw)
    if port_match:
        port = port_match.group(1).strip()
        if port and port not in ("0", "none"):
            return f"{host}:{port}"
    return host


def _resolve_componente_raw(finding: Finding, asset: Optional[Asset]) -> str:
    raw = (finding.componente_afectado or "").strip()
    if not raw and finding.raw_tool_output:
        raw = _host_from_raw(finding.raw_tool_output)
    if not raw and asset:
        raw = (asset.fqdn or asset.nombre or "").strip()
    return raw


def _clean_componente(finding: Finding, asset: Optional[Asset]) -> str:
    return preprocess_report_field(
        _resolve_componente_raw(finding, asset), strip_bullets=True, capitalize=False
    )


def _personalized_salida(componente: str, tool_output: str) -> str:
    cleaned = salida_herramienta_for_report(tool_output or "")
    if not cleaned:
        return ""
    label = componente or "N/A"
    prefix = SALIDA_PREFIX_TEMPLATE.format(componente=label)
    return f"{prefix}\n{cleaned}"


def _first_non_empty(findings: list[Finding], attr: str) -> Optional[str]:
    for finding in findings:
        value = getattr(finding, attr, None)
        if value is not None and str(value).strip():
            return str(value)
    return None


def _unique_preserve_order(values: list[str]) -> list[str]:
    seen: set[str] = set()
    result: list[str] = []
    for value in values:
        key = value.strip()
        if not key or key in seen:
            continue
        seen.add(key)
        result.append(key)
    return result


@dataclass
class GroupedReportRow:
    """Fila consolidada para una tabla Word."""

    representative: Finding
    members: list[Finding] = field(default_factory=list)
    member_ids: list[uuid.UUID] = field(default_factory=list)
    merged_componente_afectado: str = ""
    merged_raw_tool_output: str = ""


def group_findings_for_report(
    findings: list[Finding],
    assets_map: Optional[dict[uuid.UUID, Asset]] = None,
) -> list[GroupedReportRow]:
    """Agrupa hallazgos con el mismo título (normalizado) y severidad."""
    if not findings:
        return []

    assets_map = assets_map or {}
    buckets: OrderedDict[tuple[str, Severity], list[tuple[Finding, Optional[Asset]]]] = (
        OrderedDict()
    )

    for finding in findings:
        asset = assets_map.get(finding.asset_id) if finding.asset_id else None
        key = grouping_key(finding)
        buckets.setdefault(key, []).append((finding, asset))

    rows: list[GroupedReportRow] = []
    for members_with_assets in buckets.values():
        members = [f for f, _ in members_with_assets]
        components: list[str] = []
        salidas: list[str] = []

        for finding, asset in members_with_assets:
            componente = _clean_componente(finding, asset)
            if componente:
                components.append(componente)
            salida = _personalized_salida(componente, finding.raw_tool_output or "")
            if salida:
                salidas.append(salida)

        representative = members[0]
        merged_componente = "\n".join(_unique_preserve_order(components))
        merged_salida = "\n".join(salidas)

        rows.append(
            GroupedReportRow(
                representative=representative,
                members=members,
                member_ids=[f.id for f in members],
                merged_componente_afectado=merged_componente,
                merged_raw_tool_output=merged_salida,
            )
        )

    return rows


def first_non_empty_field(findings: list[Finding], attr: str) -> Optional[str]:
    return _first_non_empty(findings, attr)


SEVERITY_ORDER: dict[Severity, int] = {
    Severity.critical: 0,
    Severity.high: 1,
    Severity.medium: 2,
    Severity.low: 3,
    Severity.info: 4,
}


def severity_sort_key(finding: Finding) -> tuple[int, str]:
    """Clave de orden: severidad (CRÍTICA primero) y título A-Z sin distinguir mayúsculas."""
    sev = finding.severidad or Severity.medium
    order = SEVERITY_ORDER.get(sev, 99)
    return order, normalize_grouping_title(finding.titulo or "")


def sort_grouped_rows_by_severity(rows: list[GroupedReportRow]) -> list[GroupedReportRow]:
    """Ordena filas agrupadas: severidad descendente y título alfabético."""
    return sort_grouped_rows_for_report(rows)


def sort_grouped_rows_for_report(rows: list[GroupedReportRow]) -> list[GroupedReportRow]:
    """CRÍTICA → ALTA → MEDIA → BAJA → INFORMATIVA; dentro de cada nivel, A-Z."""
    if len(rows) <= 1:
        return rows
    return sorted(rows, key=lambda row: severity_sort_key(row.representative))


def prepare_grouped_rows_for_report(
    findings: list[Finding],
    assets_map: Optional[dict[uuid.UUID, Asset]] = None,
) -> list[GroupedReportRow]:
    """Agrupa hallazgos y devuelve filas listas para Word (ordenadas)."""
    return sort_grouped_rows_for_report(group_findings_for_report(findings, assets_map))
